assembly/src/tokens/lines.rs

```diff
@@ -0,0 +1,323 @@
+    assert_eq!(Some((1, 2)), next_token_range(" x "));
+    assert_eq!(Some((1, 4)), next_token_range(" add mul "));
+}
+
+#[test]
+fn token_lines_works_single_token() {
+    let source = r#"
+        begin
+        "#;
+    let tokens: Vec<_> = TokenLines::from(source).collect();
+    let expected = [SourceToken::new(SourceLocation::new(2, 9), "begin")].to_vec();
+    assert_eq!(tokens, expected);
+}
+
+#[test]
+fn token_lines_works_inline_tokens() {
+    let source = "begin add mul end";
+    let tokens: Vec<_> = TokenLines::from(source).collect();
+    let expected = [
+        SourceToken::new(SourceLocation::new(1, 1), "begin"),
+        SourceToken::new(SourceLocation::new(1, 7), "add"),
+        SourceToken::new(SourceLocation::new(1, 11), "mul"),
+        SourceToken::new(SourceLocation::new(1, 15), "end"),
+    ]
+    .to_vec();
+    assert_eq!(tokens, expected);
+}
+
+#[test]
+fn token_lines_works_multiline_tokens() {
+    let source = r#"begin add # foo
+            mul end"#;
+    let tokens: Vec<_> = TokenLines::from(source).collect();
+    let expected = [
+        SourceToken::new(SourceLocation::new(1, 1), "begin"),
+        SourceToken::new(SourceLocation::new(1, 7), "add"),
+        SourceToken::new(SourceLocation::new(2, 13), "mul"),
+        SourceToken::new(SourceLocation::new(2, 17), "end"),
+    ]
+    .to_vec();
+    assert_eq!(tokens, expected);
+}
+
+#[test]
+fn token_lines_works_simple_mod_comment() {
+    let source = r#"#! some mod comment      
+    begin add # foo
+
+            # bar
+
+            mul
+      end
+
+
+            # baz"#;
+    let tokens: Vec<_> = TokenLines::from(source).collect();
+    let expected = [
+        SourceToken::new(SourceLocation::new(2, 5), "begin")
+            .with_docs(["some mod comment"].to_vec()),
+        SourceToken::new(SourceLocation::new(2, 11), "add"),
+        SourceToken::new(SourceLocation::new(6, 13), "mul"),
+        SourceToken::new(SourceLocation::new(7, 7), "end"),
+    ]
+    .to_vec();
+    assert_eq!(tokens, expected);
+}
+
+#[test]
+fn token_lines_works_void_remainder_doc_comment() {
+    let source = r#"#! some mod comment      
+    begin add # foo
+
+            # bar
+
+            mul
+      end
+
+
+            #! baz"#;
+    let tokens: Vec<_> = TokenLines::from(source).collect();
+    let expected = [
+        SourceToken::new(SourceLocation::new(2, 5), "begin")
+            .with_docs(["some mod comment"].to_vec()),
+        SourceToken::new(SourceLocation::new(2, 11), "add"),
+        SourceToken::new(SourceLocation::new(6, 13), "mul"),
+        SourceToken::new(SourceLocation::new(7, 7), "end"),
+    ]
+    .to_vec();
+    assert_eq!(tokens, expected);
+}
+
+#[test]
+fn token_lines_works_inline_doc_comment() {
+    let source = r#"#! some mod comment      
+    begin add # foo
+
+            #! bar
+
+            mul
+    #!                    end doc comment with trailing spaces                
+      end
+
+
+            #! baz"#;
```

Comment:
This test should fail - doc comments must either be at the beginning of a module or be always be followed by a procedure declaration.